### PR TITLE
Add mdoc style checker

### PR DESCRIPTION
**Pull request by loli10K**
### Description

Add a new make 'mancheck' target which uses `mandoc -Tlint` to verify manpage files: currently only zfs(8), zpool(8) zdb(8) and zgenhostid(8) are supported.

Additionally fix some outstanding manpage formatting issues. (TODO)

### Motivation and Context

Hopefully avoid future formatting issues.

### How Has This Been Tested?

Run manually `mandoc -Tlint` on a Debian box.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

File: Makefile.am [hunk 1]
```diff
@@ -57,6 +57,18 @@ shellcheck:
...
 	fi
 
+mancheck:
+	@if type mandoc > /dev/null 2>&1; then \
+		for file in zfs zpool zdb zgenhostid; do \
+			mandoc -Tlint ${top_srcdir}/man/man8/$$file.8; \
+		done \
+	else \
+		type mandoc; \
+		sudo apt-get update ; sudo apt-get install -y mandoc; \
```
Comment (dinatale2) [inline Makefile.am]: I'm assuming this is temporary? I can get buildbot to install mandoc for you.

Comment (loli10K) [reply Makefile.am]: Yes, it is temporary, i was only testing why it isnt't already installed ... Brian updated to buildslaves to Ubuntu Zesty (17) in https://github.com/zfsonlinux/zfs-buildbot/commit/aa7437e7e71f7694efbacaa59e8dfc8c2f15970e but somehow it seems it's still running Xenial (16) which doesn't ship `mandoc`.

Comment (dinatale2) [reply Makefile.am]: Give me one second. I'll look into that.

Comment (dinatale2) [reply Makefile.am]: Found the problem. Will get it fixed. Thanks for catching that.

Comment (behlendorf) [reply Makefile.am]: Thanks @dinatale2.  We'll see about updating the buildbot when there's a lull in testing.

**Comment by dinatale2** [comment]
@loli10K ZFS buildbot was restarted and I reran your style build. Looks like `shellcheck` failed (I can take care of that shortly), but mandoc appears to be installed.

**Comment by loli10K** [comment]
There seems to be a bug in Ubuntu/Debian `mandoc`:

```
mandoc: ./man/man8/zfs.8:32:2: WARNING: cannot parse date, using it verbatim: June 28, 2017
[...]
mandoc: ./man/man8/zpool.8:29:2: WARNING: cannot parse date, using it verbatim: June 28, 2017
[...]
mandoc: ./man/man8/zdb.8:18:2: WARNING: cannot parse date, using it verbatim: April 14, 2017
mandoc: ./man/man8/zgenhostid.8:24:2: WARNING: cannot parse date, using it verbatim: July 24, 2017
[...]
Makefile:1331: recipe for target 'mancheck' failed
make: *** [mancheck] Error 2
```

It says it cannot parse the date, but i just built mandoc on my Debian (from jessie-backports) and the issue is at configure time: with `HAVE_STRPTIME 0` in `config.h` functions like `mandoc_normdate()/a2time()` cannot parse **any** date.

```
root@linux:~/mdocml-1.13.3# grep -r HAVE_STRPTIME
configure:HAVE_STRPTIME=
configure:#define HAVE_STRPTIME ${HAVE_STRPTIME}
mandoc.c:#if HAVE_STRPTIME
configure.local.example:HAVE_STRPTIME=0
config.h:#define HAVE_STRPTIME 0
root@linux:~/mdocml-1.13.3# gdb -q mandoc -ex 'b write' -ex 'run -Tlint /usr/src/zfs/man/man8/zgenhostid.8'
Reading symbols from mandoc...done.
Breakpoint 1 at 0x4018f0
Starting program: /root/mdocml-1.13.3/mandoc -Tlint /usr/src/zfs/man/man8/zgenhostid.8

Breakpoint 1, write () at ../sysdeps/unix/syscall-template.S:81
81	../sysdeps/unix/syscall-template.S: No such file or directory.
(gdb) bt
#0  write () at ../sysdeps/unix/syscall-template.S:81
#1  0x00007ffff7aa5423 in _IO_new_file_write (f=0x7ffff7dd7060 <_IO_2_1_stderr_>, data=0x7fffffff9f60, n=43) at fileops.c:1253
#2  0x00007ffff7aa4ae3 in new_do_write (fp=fp@entry=0x7ffff7dd7060 <_IO_2_1_stderr_>, data=data@entry=0x7fffffff9f60 "mandoc: /usr/src/zfs/man/man8/zgenhostid.8:", to_do=to_do@entry=43) at fileops.c:530
#3  0x00007ffff7aa5a36 in _IO_new_file_xsputn (f=0x7ffff7dd7060 <_IO_2_1_stderr_>, data=<optimized out>, n=43) at fileops.c:1335
#4  0x00007ffff7a7cb9d in buffered_vfprintf (s=s@entry=0x7ffff7dd7060 <_IO_2_1_stderr_>, format=format@entry=0x43e920 "%s: %s:", args=args@entry=0x7fffffffc578) at vfprintf.c:2369
#5  0x00007ffff7a7737e in _IO_vfprintf_internal (s=0x7ffff7dd7060 <_IO_2_1_stderr_>, format=0x43e920 "%s: %s:", ap=ap@entry=0x7fffffffc578) at vfprintf.c:1296
#6  0x00007ffff7a81ce7 in __fprintf (stream=<optimized out>, format=<optimized out>) at fprintf.c:32
#7  0x000000000041b286 in mmsg (t=MANDOCERR_DATE_BAD, lvl=MANDOCLEVEL_WARNING, file=0x7fffffffeeb0 "/usr/src/zfs/man/man8/zgenhostid.8", line=24, col=1, msg=0x66bf10 "July 24, 2017") at main.c:927
#8  0x000000000043448a in mandoc_msg (er=MANDOCERR_DATE_BAD, m=0x659330, ln=24, col=1, msg=0x66bf10 "July 24, 2017") at read.c:995
#9  0x00000000004324d9 in mandoc_normdate (parse=0x659330, in=0x66bf10 "July 24, 2017", ln=24, pos=1) at mandoc.c:536
#10 0x0000000000429bb5 in post_dd (mdoc=0x66bb00) at mdoc_validate.c:2136
#11 0x0000000000425e98 in mdoc_valid_post (mdoc=0x66bb00) at mdoc_validate.c:358
#12 0x0000000000423818 in rew_last (mdoc=0x66bb00, to=0x66bc30) at mdoc_macro.c:276
#13 0x00000000004239ed in rew_elem (mdoc=0x66bb00, tok=MDOC_Dd) at mdoc_macro.c:373
#14 0x0000000000425a9f in in_line_eoln (mdoc=0x66bb00, tok=MDOC_Dd, line=24, ppos=1, pos=0x7fffffffc870, buf=0x669470 ".Dd July") at mdoc_macro.c:1403
#15 0x0000000000421188 in mdoc_macro (mdoc=0x66bb00, tok=MDOC_Dd, line=24, ppos=1, pos=0x7fffffffc870, buf=0x669470 ".Dd July") at mdoc.c:274
#16 0x000000000042228c in mdoc_pmacro (mdoc=0x66bb00, ln=24, buf=0x669470 ".Dd July", offs=17) at mdoc.c:805
#17 0x0000000000420fe5 in mdoc_parseln (mdoc=0x66bb00, ln=24, buf=0x669470 ".Dd July", offs=1) at mdoc.c:247
#18 0x00000000004337f1 in mparse_buf_r (curp=0x659330, blk=..., i=934, start=1) at read.c:600
#19 0x0000000000433bc2 in mparse_parse_buffer (curp=0x659330, blk=..., file=0x7fffffffeeb0 "/usr/src/zfs/man/man8/zgenhostid.8") at read.c:741
#20 0x0000000000433cb4 in mparse_readfd (curp=0x659330, fd=3, file=0x7fffffffeeb0 "/usr/src/zfs/man/man8/zgenhostid.8") at read.c:778
#21 0x000000000041a8cd in parse (curp=0x7fffffffcb00, fd=3, file=0x7fffffffeeb0 "/usr/src/zfs/man/man8/zgenhostid.8", level=0x7fffffffcabc) at main.c:644
#22 0x000000000041a050 in main (argc=1, argv=0x7fffffffeca8) at main.c:437
(gdb) list a2time
461		return(start);
462	}
463	
464	static int
465	a2time(time_t *t, const char *fmt, const char *p)
466	{
467		struct tm	 tm;
468		char		*pp;
469	
470		memset(&tm, 0, sizeof(struct tm));
(gdb) 
471	
472		pp = NULL;
473	#if HAVE_STRPTIME
474		pp = strptime(p, fmt, &tm);
475	#endif
476		if (NULL != pp && '\0' == *pp) {
477			*t = mktime(&tm);
478			return(1);
479		}
480	
```

Sorry but this PR/feature won't work until this gets fixed in Ubuntu, unless we start building mandoc from source on the STYLE buildslaves.

**Comment by behlendorf** [comment]
Until we can get Ubuntu/Debian to fix this bug I believe we can force to only report errors and suppress the warnings, ` mandoc -Tlint -Werror ...`.  Not ideal, but better than no checking.

**Comment by loli10K** [comment]
Agreed, done.

**Pull request merged**